mls_validation_service/src/cached_signature_verifier.rs

```diff
@@ -0,0 +1,187 @@
+use lru::LruCache;
+use std::num::NonZeroUsize;
+use tokio::sync::Mutex;
+
+use ethers::types::{BlockNumber, Bytes};
+use xmtp_id::associations::AccountId;
+use xmtp_id::scw_verifier::{SmartContractSignatureVerifier, ValidationResponse, VerifierError};
+
+#[derive(Hash, Eq, PartialEq, Clone, Debug)]
+pub struct CacheKey {
+    pub address: String,
+    pub chain_id: String,
+    pub hash: [u8; 32],
+    pub signature: Vec<u8>,
+    pub block_number: Option<u64>,
+}
+
+impl CacheKey {
+    pub fn new(
+        account_id: &AccountId,
+        hash: [u8; 32],
+        signature: &Bytes,
+        block_number: Option<BlockNumber>,
+    ) -> Self {
+        let block_number_u64 = block_number.and_then(|bn| bn.as_number().map(|n| n.as_u64()));
+        let address = account_id.get_account_address().to_string();
+        let chain_id = account_id.get_chain_id().to_string();
+
+        Self {
+            chain_id,
+            address,
+            hash,
+            signature: signature.to_vec(),
+            block_number: block_number_u64,
+        }
+    }
+}
+
+/// A cached smart contract verifier.
+///
+/// This wraps MultiSmartContractSignatureVerifier (or any other verifier
+/// implementing SmartContractSignatureVerifier) and adds an in-memory LRU cache.
+pub struct CachedSmartContractSignatureVerifier {
+    verifier: Box<dyn SmartContractSignatureVerifier>,
+    cache: Mutex<LruCache<CacheKey, ValidationResponse>>,
+}
+
+impl CachedSmartContractSignatureVerifier {
+    pub fn new(
+        verifier: impl SmartContractSignatureVerifier + 'static,
+        cache_size: usize,
+    ) -> Result<Self, VerifierError> {
+        if cache_size == 0 {
+            return Err(VerifierError::InvalidCacheSize(cache_size));
```

Comment:
Rather than creating this special error case, you can change `cache_size` to be a `NonZeroUsize`